grin/src/AbstractInterpretation/CreatedBy.hs

```diff
@@ -0,0 +1,278 @@
+              bindM
+              altM >>= \case
+                Z -> emit IR.Set {dstReg = caseResultReg, constant = IR.CSimpleType unitType}   -- pure ()
+                R altResultReg -> emit IR.Move {srcReg = altResultReg, dstReg = caseResultReg}
+
+        case cpat of
+          NodePat tag vars -> do
+            irTag <- getTag tag
+            altInstructions <- codeGenAlt $ do
+              -- restrict scrutinee to alternative's domain
+              flip (maybe (pure ())) scrutRegMapping $ \(name, _) -> do
+                altScrutReg <- newReg
+                addReg name altScrutReg
+                -- NOTE: We just create a new empty register, and associate it with the scrutinee in this alternative. Then we annotate the register with restricted properties of the scrutinee.
+                emit IR.Project
+                  { srcSelector = IR.ConditionAsSelector $ IR.NodeTypeExists irTag
+                  , srcReg = valReg
+                  , dstReg = altScrutReg
+                  }
+
+              -- bind pattern variables
+              forM_ (zip [1..] vars) $ \(idx, name) -> do
+                  argReg <- newReg
+                  addReg name argReg
+                  emit IR.Project {srcSelector = IR.NodeItem irTag idx, srcReg = valReg, dstReg = argReg}
+            emit IR.If {condition = IR.NodeTypeExists irTag, srcReg = valReg, instructions = altInstructions}
+
+          LitPat lit -> do
+            altInstructions <- codeGenAlt $
+              -- restrict scrutinee to alternative's domain
+              flip (maybe (pure ())) scrutRegMapping $ \(name, _) -> do
+                altScrutReg <- newReg
+                addReg name altScrutReg
+                emit IR.Project
+                  { srcSelector = IR.ConditionAsSelector $ IR.SimpleTypeExists (litToSimpleType lit)
+                  , srcReg = valReg
+                  , dstReg = altScrutReg
+                  }
+            emit IR.If {condition = IR.SimpleTypeExists (litToSimpleType lit), srcReg = valReg, instructions = altInstructions}
+
+          DefaultPat -> do
+            tags <- Set.fromList <$> sequence [getTag tag | A (NodePat tag _) _ <- alts]
+            altInstructions <- codeGenAlt $
+              -- restrict scrutinee to alternative's domain
+              flip (maybe (pure ())) scrutRegMapping $ \(name, _) -> do
+                altScrutReg <- newReg
+                addReg name altScrutReg
+                emit IR.Project
+                  { srcSelector = IR.ConditionAsSelector $ IR.NotIn tags
+                  , srcReg = valReg
+                  , dstReg = altScrutReg
+                  }
+            emit IR.If {condition = IR.NotIn tags, srcReg = valReg, instructions = altInstructions}
+
+          _ -> throwCBy $ "CBy does not support the following case pattern: " ++ show cpat
+
+      -- restore scrutinee register mapping
+      maybe (pure ()) (uncurry addReg) scrutRegMapping
```

Comment:
You could use `forM_` instead of `maybe (pure ())`.

Reply:
Done.